### DISCUSSION TITLE
Feature Request: KeyboardSelection Behavior or AttachedProperty

### DISCUSSION THREAD

**Discussion by sonnemaf** [CommunityToolkit/Labs-Windows#251] [Ideas]
## I'm submitting a...
Feature request for a KeyboardSelection Behavior or AttachedProperty

UserVoice request: https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/36293008-keyboardselection-in-selectors-listbox-combobox

## Current behavior
Keyboard selection in Selectors (ListBox, ComboBox, ListView, GridView) for UWP apps doesn't really work.

Suppose you have a ListBox which is databound to a employees collection. The DataTemplate of the ListBox show the Name and Salary of the employee in a StackPanel. If the ListBox has the focus and you hit for example the key S you would expect the selection of the ListBox to go to the first employee which name starts with the letter S. This doesn't happen, nothing happens.

```
<ListBox ItemsSource="{Binding Employees}">
    <ListBox.ItemTemplate>
        <DataTemplate>
            <StackPanel>
                    <TextBlock Text="{Binding Name}" FontWeight="Bold"/>
                    <TextBlock Text="{Binding Salary}"/>
            </StackPanel>
        </DataTemplate>
    </ListBox.ItemTemplate>
</ListBox>
```

## Expected behavior
With a KeyboarSelectionBehavior you could add the behavior (select the correct item in the listbox using a keyboard). It can aslo be implemented using an AttachedProperty but I think a Behavior makes it more discoverable when using Blend.

```
<ListBox ItemsSource="{Binding Employees}">
    <i:Interaction.Behaviors>
        <local:KeyboardSelectionBehavior SelectionMemberPath="Name"/>
    </i:Interaction.Behaviors>
    <ListBox.ItemTemplate>
        <DataTemplate>
            <StackPanel>
                <TextBlock Text="{Binding Name}" FontWeight="Bold"/>
                <TextBlock Text="{Binding Salary}"/>
            </StackPanel>
        </DataTemplate>
    </ListBox.ItemTemplate>
</ListBox>
```

I have already written 2 implementations (WinRT and Silverlight):
https://reflectionit.nl/Blog/2013/windows-8-xaml-tips-creating-blend-behaviors 
https://www.reflectionit.nl/blog/2010/keyboard-selection-on-silverlight-listbox-and-combobox

I would love to implement this feature myself if you find it useful.

**Comment by Kyaa-dost** [comment]
@sonnemaf I know it's been a while and thanks for suggesting this feature. Are you still willing to work on this feature if the community finds it suitable?

**Comment by sonnemaf** [comment]
Yes I am sure.

**Comment by michael-hawker** [comment]
@sonnemaf I'm trying to get some metrics to understand if we should have a dedicated behavior package. If so, I'll split that out next week and let you know, so all these related behavior items you're working on can go there!

**Comment by michael-hawker** [comment]
@sonnemaf I know your Voice Command PR got stuck as we had trouble understanding how permissions were behaving with existing apps (seems like a system bug we should diagnose). I think it's something we can move to our upcoming Labs environment at least to get more data from other folks with as we try and find the right folks to talk to about how those permissions should work when upgrading an app.

In the meantime, would you want to work on your other behaviors for a future version instead?

**Comment by sonnemaf** [comment]
Hi @michael-hawker,

I would love to start on the development of some new Behaviors in October. September is very busy. Is that ok?

It is ok to move the  Voice Command PR to the new Labs environment. Not sure what the Labs environment is/will-be. I will learn it on the way.

Fons

**Comment by michael-hawker** [comment]
Sounds great @sonnemaf, thanks! I'll be working on setting up Labs hopefully in September, so think our timings sound all aligned. No rush! Appreciate your continued enthusiasm! 🦙❤
